realtek: rtl839x: keep vlan tags on forwarded pkts

luizluca opened the pull request:
rtl839x was dropping vlan tags on packets forwarded from a user port to another one. Packets from the cpu were not affected.

The same BIT(4) is used by the factory image (read using /dev/mem).

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

luizluca committed:
realtek: refactor keep vlan tag setup, fix tagged forwarding

The code in dsa.c:rtl83xx_port_enable() was trying to set
vlan_port_tag_sts_ctrl while dealing with differences between SoCs.
However, not only that register has a different address, the register
structure and even the 2-bit value semantic changes for each SoC.

The vlan_port_tag_sts_ctrl field was dropped and converted into a
vlan_port_keep_incoming_tag_set() function that abstracts the different
between SoCs. The macro referencing that register migrated to the SoC
specific c file as it will be privately used by each file.

All magic numbers were converted into macros using BITMASK and
FIELD_PREP.

The vlan_port_tag_sts_ctrl debugfs was dropped for now as it is already
broken for rtl93xx. The best place for SoC specific code might be in each
respective c file and not in if/else clauses.

The final result is:

rtl838x: set ITAG_STS=TAGGED, same as before
rtl839x: set ITAG_STS=TAGGED instead of IGR_P_ITAG_KEEP=0x1, fixing
	 forwarding of tagged packets
rtl930x: set EGR_ITAG_STS=TAGGED instead of IGR_P_ITAG=0x1, possibly
	 fixing forwarding of tagged packets
rtl931x: set EGR_ITAG_STS=TAGGED instead of OTPID_KEEP=0x1, possibly
         fixing forwarding of tagged packets

Without (EGR_)ITAG_STS=TAGGED, at least for rtl839x, forwarded packets
will drop the vlan tag while packets from the CPU will still have the
correct tag.

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

luizluca commented:
FYI https://forum.openwrt.org/t/support-for-rtl838x-based-managed-switches/57875/2174

luizluca commented:
It should probably be backported to 22.03 as rtl8393_zyxel_gs1900-48.dts is affected as well (I don't have that switch to confirm)

svanheule commented:
Thanks for looking into this.

It looks like rtl839x and rtl930x are similar, rtl838x was correct, and rtl931x is wrong:
* maple/RTL839x: https://svanheule.net/realtek/maple/register/vlan_port_tag_sts_ctrl
* cypress/RTL839x: https://svanheule.net/realtek/cypress/register/vlan_port_tag_sts_ctrl
* longan/RTL930x: https://svanheule.net/realtek/longan/register/vlan_port_tag_sts_ctrl
* mango/RTL931x: https://svanheule.net/realtek/mango/register/vlan_port_tag_ctrl

Until I can give this a proper look, here's some extra info I was able to find in the SDK:
```
$ rg -B1 -i ITAG_STS
include/hal/mac/mac_debug.h
1090-            uint32 ale_otag_sts:2;      /*ale outer tag status*/
1091:            uint32 ale_itag_sts:2;      /*ale inner tag status*/
--
2767-            uint32 otag_sts:2;      /*00: keep, 01:untag, 10:tag, 11: ale no determined*/
2768:            uint32 itag_sts:2;      /*00: keep, 01:untag, 10:tag, 11: ale no determined*/

src/dal/maple/dal_maple_acl.c
7607-            case 0:
7608:                pAction->tag_sts_data.itag_sts = ACL_ACTION_TAG_STS_UNTAG;
--
7610-            case 1:
7611:                pAction->tag_sts_data.itag_sts = ACL_ACTION_TAG_STS_TAG;
--
7613-            case 2:
7614:                pAction->tag_sts_data.itag_sts = ACL_ACTION_TAG_STS_KEEP_CONTENT;
--
7616-            case 3:
7617:                pAction->tag_sts_data.itag_sts = ACL_ACTION_TAG_STS_NOP;

```

If the commit message and code properly document the reasoning behind the change, I'm OK with this patch changing all platform to use the `ITAG_STS` field consistently across platforms. You've tested RTL839x, we know RTL838x works, so I think the other two are likely to work as well.

plappermaul commented:
Good catch and thanks for the hard work. Nothing to add here or to complain about.

piaddic7ed commented:
@luizluca thanks for fixing this issue.
@oliv3r can you please pick this for your current development branch? I'd like to test this with my rtl930x switch.

luizluca commented:
> @luizluca thanks for fixing this issue.
> @oliv3r can you please pick this for your current development branch? I'd like to test this with my rtl930x switch.

I believe it does not change the behavior for rtl930x. However, if your device is also affected by a similar issue, you might want to add it to the switch (if they share the same register structure.

oliv3r commented:
> @luizluca thanks for fixing this issue. @oliv3r can you please pick this for your current development branch? I'd like to test this with my rtl930x switch.

Sure can do; though my branch is foobar atm :(

but as @luizluca  states that this doesn't change rtl930x I'm don't mind waiting for this commit to land via master?

luizluca commented:
> Thanks for looking into this.
> 
> It looks like rtl839x and rtl930x are similar, rtl838x was correct, and rtl931x is wrong:

Should I add rtl931x? Maybe commented?

> * maple/RTL839x: https://svanheule.net/realtek/maple/register/vlan_port_tag_sts_ctrl
> * cypress/RTL839x: https://svanheule.net/realtek/cypress/register/vlan_port_tag_sts_ctrl
> * longan/RTL930x: https://svanheule.net/realtek/longan/register/vlan_port_tag_sts_ctrl
> * mango/RTL931x: https://svanheule.net/realtek/mango/register/vlan_port_tag_ctrl

This is nice. Thanks.

> If the commit message and code properly document the reasoning behind the change, I'm OK with this patch changing all platform to use the `ITAG_STS` field consistently across platforms. You've tested RTL839x, we know RTL838x works, so I think the other two are likely to work as well.

I'll try to improve the commit and add a case for every know SoC as those references will help highlight where SoCs differs.

Maybe it would be better to have a func that setups that register for each SoC instead of a pointer to related registers. They are simply different registers with equal/similar names/functions but different usage.

svanheule commented:
RTL931x is the only exception in the current code, which results in the following behaviour:
* rtl838x: set `ITAG_STS` to 0x1
* rtl839x: set `IGR_P_ITAG_KEEP` to 0x1
* rtl930x: set `IGR_ITAG_KEEP` to 0x1
* rtl931x: set `OTPID_IDX` to 0x1

To fix tagging, your patch changes this to:
* rtl839x: set `ITAG_STS` to 0x1

The excerpts I quoted are probably not relevant, as they relate to the ACL tables. From XGS1250/sdk/src/dal/cypress/dal_cypress_vlan.c (dal_maple_vlan.c has similar code):
```c
    if ((ret = reg_array_field_read(unit, CYPRESS_VLAN_PORT_TAG_STS_CTRLr, port, REG_ARRAY_INDEX_NONE,\
                CYPRESS_ITAG_STSf, &value)) != RT_ERR_OK)
        [...]

    /* translate chip's value to definition */
    switch (value)
    {
        case 0:
            *pSts = TAG_STATUS_UNTAG;
            break;
        case 1:
            *pSts = TAG_STATUS_TAGGED;
            break;
        case 2:
            *pSts = TAG_STATUS_PRIORITY_TAGGED;
            break;
        default:
            return RT_ERR_FAILED;
    }
```

It also appears that for `EGR_ITAG_STS` on rtl930x and rtl931x, the meaning differs:
```c
    /* translate chip's value to definition */
    switch (value)
    {
        case 0:
            *pSts = TAG_STATUS_INTERNAL;
            break;
        case 1:
            *pSts = TAG_STATUS_UNTAG;
            break;
        case 2:
            *pSts = TAG_STATUS_TAGGED;
            break;
        case 3:
            *pSts = TAG_STATUS_PRIORITY_TAGGED;
            break;
        default:
            return RT_ERR_FAILED;
    }
```

I suspect the comment (from our driver) `/* enable inner tagging on egress, do not keep any tags */` refers to clearing the `*_KEEP` fields, while setting the `ITAG_STS` fields to the correct value.

For RTL930x, I'm not sure what this means for the current driver state. It _might_ be that VLAN tagging (mostly) works. If so, there may be less common cases (e.g. sending a frame with a wrong VLAN tag to a certain port) where it fails, but I can't currently test this.

Unless somebody can test VLANs with RTL930x (on short notice), perhaps we should keep the current behaviour for that platform and RTL931x, but mark the code as highly suspicious. Platform-specific setup calls also sounds like a good idea. There could still be small (inline) functions for the shared parts, although that's maybe not worth the effort if it only gets used twice.

luizluca commented:
Thanks @svanheule ,

Based on the devices I have here rtl838x/rtl839x and on what you shared, I updated the PR with a more radical path.
I still did not test it (tomorrow I will), but it seems to be a better approach for the issue. It would be really nice if we could have some rtl930x/rtl931x users to test current state and after the patch was applied.

I believe vlan_port_keep_tag_set() signature might not be ideal but it is, at least, enough for a first step.

I dropped the reference for that register from debugfs as that code would also need some love. I believe most `if (family) {} else {}` should become functions. If not, I would change them to switch/case, with the default case being an error. It is too easy to miss a SoC peculiarity when `if (family) {} else {}` is in use.

luizluca commented:
> I still did not test it (tomorrow I will),

Tested and it work as expected. It would be nice to reproduce de problem with rtl930x and rtl931x and check if the same issue happens there (and if this patch fixes it).

@piaddic7ed, now the patch affects rtl930x

oliv3r commented on target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/rtl930x.c:
```diff
@@ -5,6 +5,22 @@
+		RTL930X_VLAN_PORT_TAG_STS_CTRL_BASE + (port << 2)
+#define RTL930X_VLAN_PORT_TAG_STS_CTRL_EGR_OTAG_STS_MASK	GENMASK(7,6)
+#define RTL930X_VLAN_PORT_TAG_STS_CTRL_EGR_ITAG_STS_MASK	GENMASK(5,4)
+#define RTL930X_VLAN_PORT_TAG_STS_CTRL_EGR_P_OTAG_KEEP_MASK	GENMASK(3,3)
```
I think you meant to say `BIT(3)` here :)

luizluca replied:
Aren't they equivalent? For consistency with other masks, I would keep GENMASK as it is just a coincidence that the mask is a single bit.

piaddic7ed commented:
> @piaddic7ed, now the patch affects rtl930x

Well, I do have a xgs1010-12 which is not yet supported. However, booting the 1250 initramfs mostly works. What exactly is supposed to have improved? Tagged ports are not something I can test at the moment. I cannot see any improvement in my setup at least. In my case packets to and from the CPU port don't seem to consistently work.
My setup is as follows:
Vlan 1 (default) on port 1, interface ip 192.168.1.1, client connected to port1 (192.168.1.111).
Vlan 100 on port 2, interface ip 192.168.100.1, client connected to port1 (192.168.100.111).
Vlan 200 on port 3, interface ip 192.168.200.1, client connected to port1 (192.168.200.111).
Clients in all networks can ping 192.168.1.1 and 192.168.200.1, but none can ping 192.168.100.1.
Form the switch similar picture, I can ping both the .1 and .200 clients, but not the .100 client.
Client to client communication is broken in similar ways.

On rtl838x this setup works fine.

luizluca commented:
> Well, I do have a xgs1010-12 which is not yet supported. However, booting the 1250 initramfs mostly works. What exactly is supposed to have improved? Tagged ports are not something I can test at the moment.

This patch is all about tags. Consider the setup:

Port 1: untagged VLAN 10, tagged VLAN 20
Port 2: untagged VLAN 10, tagged VLAN 20

Client 1 connected to port 1, with 192.168.10.11 on untagged and 192.168.20.11 on a slave VLAN interface

Client 2 connected to port 2, with 192.168.10.12 on untagged and 192.168.20.12 on a slave VLAN interface.

Before this patch, client 1 can ping 192.168.10.12 but not 192.168.20.12 as the packet that should have the VLAN 20 tag will arrive at client 2 without any tag. After the patch, that packet will have the VLAN 20 tag.

I cannot see any improvement in my setup at least. In my case packets to and from the CPU port don't seem to consistently work.
> My setup is as follows:
> Vlan 1 (default) on port 1, interface ip 192.168.1.1, client connected to port 1 (192.168.1.111).
> Vlan 100 on port 2, interface ip 192.168.100.1, client connected to port 2 (192.168.100.111).
> Vlan 200 on port 3, interface ip 192.168.200.1, client connected to port 3 (192.168.200.111).
> Clients in all networks can ping 192.168.1.1 and 192.168.200.1, but none can ping 192.168.100.1.
> Form the switch similar picture, I can ping both the .1 and .200 clients, but not the .100 client.
> Client to client communication is broken in similar ways.

It seems thst port 2 or VLAN 100 is not working. But this is not related to this PR. It is better to discuss this on a dedicated issue.

luizluca commented:
> Thanks @svanheule ,
> 
> Based on the devices I have here rtl838x/rtl839x and on what you shared, I updated the PR with a more radical path. I still did not test it (tomorrow I will), but it seems to be a better approach for the issue. It would be really nice if we could have some rtl930x/rtl931x users to test current state and after the patch was applied.
> 
> I believe vlan_port_keep_tag_set() signature might not be ideal but it is, at least, enough for a first step.
> 
> I dropped the reference for that register from debugfs as that code would also need some love. I believe most `if (family) {} else {}` should become functions. If not, I would change them to switch/case, with the default case being an error. It is too easy to miss a SoC peculiarity when `if (family) {} else {}` is in use.

Did I go too far? ACK/NACK? At least it fixes broken shipped code (22.03.x).

svanheule commented:
The new changes seemed fine when I quickly went over them, but I was actually waiting for a clearer test result on RTL930x. I moved house recently and have been busy with other things. My switches are still in boxes, so I haven't been able to test things myself. I'll try to find some time to gather up the Realtek fixes that have been pending for a few weeks now, but can't make any hard promises on a timeline.

svanheule commented:
Merged to my staging tree. Let's see if anybody screams about RTL930x breaking.

For 22.03, would it be an option to reduce this patch, and instead only put the fixed values for the registers (for rtl839x and maybe rtl930x) in dsa.c? Perhaps something that can be discussed over that patch.

luizluca commented:
We could try a smaller patch but anything smaller might require extra if and magic bit ops. I can try something after 12/12 as I'm away from office

svanheule commented:
That's true, and keeping the patch the same also reduces the chances of things (unintentionally) breaking in a different way. Let's try that then, and just see if it applies cleanly to 22.03 or not.
